src/util/CopyableUniquePtr.h

```diff
@@ -0,0 +1,92 @@
+// Copyright 2023, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Andre Schlegel (April of 2023, schlegea@informatik.uni-freiburg.de)
+
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include "util/json.h"
+
+namespace ad_utility {
+
+/*
+A version of `std::unique_ptr` with a copy assigment operator and a copy
+constructor, which both copy the dereferenced pointer to create a new instace
+of the object for the `unique_prt`.
+Currently not written with support for dynamically-allocated array of objects
+in mind, so that may not work.
+*/
+template <typename T, typename Deleter = std::default_delete<T>>
+requires std::is_copy_constructible_v<T>
+class CopyableUniquePtr : public std::unique_ptr<T, Deleter> {
+  // This makes calling functions, etc. from the base class so much easier.
+  using Base = std::unique_ptr<T, Deleter>;
+
+  // This function uses a private constructor, so it needs private access.
+  template <typename T2, typename... Args>
+  friend constexpr CopyableUniquePtr<T2> make_copyable_unique(Args&&... args);
+
+  /*
+  @brief Returns an unique pointer, that holds a copy of the dereferenced
+  (copyable) unique pointer, that was given.
+  */
+  Base CopyDereferencedPointer(const auto& ptr) {
+    // Different behaviour based on whenever the ptr actually owns an object.
+    return ptr ? std::make_unique<T>(*ptr) : nullptr;
+  }
+
+  /*
+  @brief Creates a `CopyableUniquePtr`, that holds the object formerly
+  owned by the unique pointer. Needed for `make_copyable_unique`.
+  */
+  explicit CopyableUniquePtr(Base&& ptr) : Base(std::move(ptr)) {}
```

Comment:
Order:
first data members and types,
then public methods, then private methods.

Reply:
Where did you get this order from? I've never heard of it.

Reply:
Done, see [5d680fd](https://github.com/ad-freiburg/qlever/pull/860/commits/5d680fdff2709889008f979020d2f4eac3fc465f).